contrib/codegen/src/database.rs

```diff
@@ -0,0 +1,144 @@
+    let invocation = parse_invocation(attr, input)?;
+
+    if let Type::Path(type_path) = invocation.connection_type {
+        let inner_type = &type_path.path.segments.iter()
+            .map(|ps| ps.ident.clone())
+            .collect::<Vec<Ident>>();
+
+        let database_name = &invocation.db_name;
+        let handler_type = &invocation.type_name;
+        let pool_type = Ident::new(&format!("{}Pool", handler_type), Span::call_site());
+
+        let tokens = quote! {
+            use ::std::ops::*;
+            use ::rocket::http::*;
+            use ::rocket::request::*;
+            use ::rocket::*;
+            use ::rocket::fairing::*;
+            use ::rocket_contrib::*;
+            use ::rocket_contrib::databases::*;
+            pub struct #handler_type(r2d2::PooledConnection<<#(#inner_type)::* as Poolable>::Manager>);
+            pub struct #pool_type(r2d2::Pool<<#(#inner_type)::* as Poolable>::Manager>);
+
+            impl #handler_type {
+                pub fn pool(config: DatabaseConfig) -> Result<#pool_type, <#(#inner_type)::* as Poolable>::Error> {
+                    <#(#inner_type)::* as Poolable>::manager(&config)
+                        .and_then(|manager| {
+                             match r2d2::Pool::builder().max_size(config.pool_size).build(manager) {
+                                 Ok(pool) => Ok(pool),
+                                 Err(e) => Err(databases::DbError::PoolError(e)) // TODO: Fix this
+                             }
+                        })
+                        .and_then(|pool| Ok(#pool_type(pool)))
+                }
+
+                pub fn fairing() -> impl Fairing {
+                    AdHoc::on_attach(move |rocket| {
+                        if let Some(database_config) = databases::database_config(
+                            #database_name,
+                            &rocket.config().clone()
+                        ) {
+                            return match Self::pool(database_config) {
+                                Ok(pool) => Ok(rocket.manage(pool)),
+                                Err(_) => Err(rocket)
+                            }
+                        } else {
+                            error_!("Could not find database: {}", #database_name);
+                        }
+
+                        Err(rocket)
+                    })
+                }
+            }
+
+            impl Deref for #handler_type {
+                type Target = #(#inner_type)::*;
+
+                fn deref(&self) -> &Self::Target {
+                    &self.0
+                }
+            }
+
+            impl<'a, 'r> FromRequest<'a, 'r> for #handler_type {
+                type Error = ();
+
+                fn from_request(request: &'a Request<'r>) -> request::Outcome<Self, Self::Error> {
+                    let pool = request.guard::<State<#pool_type>>()?;
+
+                    match pool.0.get() {
```

Comment:
Can this logic be provided by a public method on the pool type? I would like to be able to do initialization before launch:

```rust
.attach(DbConnection::fairing())
.attach(AdHoc::on_attach(|rocket| {
    let pool: &DbConnectionPool = rocket.state().expect("database connection pool");
    // there is no Request yet, but FromRequest is currently the only way to get a connection from the pool
    if let Ok(conn) = pool.get() {
         // run some database initialization
    }
    // return success or failure
}))
```

This can be implemented in downstream crates with the following code, but it feels wrong to do so because it relies on the implementation details of `*Pool`.
```rust
impl DbConnectionPool {
    pub fn get(&self) -> Result<DbConnection, r2d2::Error> {
        self.0.get().map(DbConn)
    }
}
```

Reply:
Hm... interesting desire in usage with the connection. There's no real reason that logic can't be extracted out and just used by the `FromRequest` trait.